hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+        """
+        Looks up the bot id by name from the users list, then memoizes it.
+        """
+
+        # Look up the bot id from the users list
+        resp = self.slack.api_call("users.list")
+        if resp.get("ok"):
+            # Retrieve the users and find the bot
+            for member in resp.get("members"):
+                if 'name' in member and member.get('name') == self.name:
+                    return member.get('id')
+
+        # If we get this far without a botid, we have a problem.
+        raise SlackException(
+            "could not find an ID for a bot named {}".format(self.name)
+        )
+
+    @memoized
+    def atbotid(self):
+        """
+        Computes and memoizes the @botid regular expression for directed
+        messages that may contain commands. E.g. the pattern for <@botid>.
+        """
+        return re.compile(
+            re.escape("<@{}>".format(self.botid)), re.I
+        )
+
+    def run(self):
+        """
+        Connects to the Slack real-time messaging API, polling every second
+        for messages
+        """
+        # Set up the signal handlers
+        # TODO: use threads to better manage communication
+        self.shutdown = False
+        signal.signal(signal.SIGINT, lambda signal, frame: self.stop())
+
+        # Fetch and cache the current weather at startup
+        self.weather()
+
+        # Schedule timed events
+        self.set_schedule()
+
+        # Connect to the real time message API
+        if self.slack.rtm_connect():
+            # TODO: add better chat logging
+            print("slackbot named {} ({}) is connected".format(self.name, self.botid))
+
+            while True:
+
+                # Check if we're shutdown
+                if self.shutdown:
+                    break
+
+                # Check to see if anything has come through the channel
+                self.read_rtm_channel()
+
+                # Run any scheduled tasks
+                schedule.run_pending()
+
+                # Sleep for a bit until the next time we read the channel
+                time.sleep(POLLING_INTERVAL)
+
+        else:
+            raise SlackException(
+                "could not connect to API: invalid Slack API key or Bot ID?"
+            )
+
+        # If we've made it here, we've been shutdown
+        # TODO: convert to logging functionality
+        print("bot has been gracefully shutdown")
+
+    def stop(self):
+        """
+        Set the shutdown semaphore
+        """
+        self.shutdown = True
+
+    def read_rtm_channel(self):
+        """
+        Reads the RTM channel, parses output, filters messages based on type
+        and if they are directed at the bot. If so, passes those messages on
+        to the message handler command.
+        """
+
+        # Check for messages on the wire
+        # TODO: as soon as we connect, the last message comes through.
+        messages = self.slack.rtm_read()
+        if messages and len(messages) > 0:
```

Comment:
Yes, and actually, I disliked that line of code very much, it is direct from the [Slackbot tutorial](https://www.fullstackpython.com/blog/build-first-slack-bot-python.html) I used (which is cited at the top of the bot.py file). The conditional `if messages` should be sufficient. 

However, at that point, I did not know what the `messages` data structure was. When you call `len(obj)`, it actually calls `obj.__len__()`, so if messages was not a list then the `len(messages)` may have performed an API call or did some parsing of underlying XML or something like that. Since I didn't know, I left it. Later, I found out that it was in fact a Python list, so that statement could be removed. 

More generally, about `pylint` or other linters/style enforcement -- my personal belief that code is about communication, both to other developers and future you. This means that you can express yourself, and have a style. In fact, if you saw my mentor, Peter Eacmen's code you would probably recognize it and my style of code. It also means that you need to follow basic principles of understandability that go beyond syntax. 

Therefore for me, a minimalistic linter is critical to success at programming and collaboration. In Python, tabs vs. spaces is critical to the parsing, so one must be decided on and it's better if the community does it rather than individual teams. In Go, `go fmt` standardizes code bases to optimize the compiler. However, I have experienced linters that are probably too strict, that go beyond simply easing communication towards cramming programs into a box of unit abstraction, which doesn't do anyone any good.